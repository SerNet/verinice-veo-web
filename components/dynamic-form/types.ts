/*
 * verinice.veo web
 * Copyright (C) 2022  Jonas Heitmann
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
import { JSONSchema7 } from 'json-schema';

export default {};

export interface IVeoFormsAdditionalContext {
  [pointer: string]: {
    objectSchema?: Record<string, any>;
    formSchema?: Record<string, any>;
  };
}

export interface IVeoFormsReactiveFormActions {
  [pointer: string]: ((
    newValue: any,
    oldValue: any,
    newObject: Record<string, any>,
    oldObject: Record<string, any>
  ) => Record<string, any>)[];
}

export interface IVeoFormsTranslations {
  [lang: string]: Record<string, any>;
}

export interface IVeoFormElementRule {
  visible: boolean;
  disabled: boolean;
}

export type IDynamicFormElementOptions = Record<string, any> &
  IVeoFormElementRule;

export interface IVeoFormElementDefaultProps {
  metaData: JSONSchema7;
  disabled: boolean;
  objectCreationDisabled: boolean;
  debug: boolean;
  [key: string]: any;
}

export interface IVeoFormsElementDefinition {
  code: string;
  name: { [lang: string]: string }; // Name of this component (used in FormSchemaEditor)
  description: { [lang: string]: string }; // Description describing this component (used in FormSchemaEditor)
  conditions?: (_props: IVeoFormElementDefaultProps) => boolean[]; // Every condition has to be truthy in order for this component to get displayed. Components with more truthy conditions will be displayed first.
  bias?: number; // Bias so this component gets choosen over other components with the same amount of truthy conditions
}

export interface IVeoFormElementFormSchemaRule {
  effect: 'HIDE' | 'SHOW' | 'DISABLE' | 'ENABLE';
  condition: {
    scope: string;
    schema: JSONSchema7;
  };
}

export interface IVeoFormElementFormSchema {
  type: string;
  options: IDynamicFormElementOptions;
  elements?: IVeoFormElementFormSchema[];
  rule?: IVeoFormElementFormSchemaRule;
}

export interface IVeoFormControlFormSchema extends IVeoFormElementFormSchema {
  scope: string;
  label: string;
}

export interface IVeoFormLabelFormSchema extends IVeoFormElementFormSchema {
  text: string;
}

export interface IVeoFormWidgetFormSchema extends IVeoFormElementFormSchema {
  name: string;
}
