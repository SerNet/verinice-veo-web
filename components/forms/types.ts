/*
 * verinice.veo web
 * Copyright (C) 2022  Jonas Heitmann
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
import { JSONSchema7 } from 'json-schema';
import { IBaseObject } from '~/lib/utils';

export default {};

export interface IVeoFormsAdditionalContext {
  [pointer: string]: {
    objectSchema?: IBaseObject;
    formSchema?: IBaseObject;
  };
}

export interface IVeoFormsReactiveFormActions {
  [pointer: string]: ((newValue: any, oldValue: any, newObject: IBaseObject, oldObject: IBaseObject) => IBaseObject)[];
}

export interface IVeoFormsTranslations {
  [lang: string]: IBaseObject;
}

export interface IVeoFormElementRule {
  visible: boolean;
  disabled: boolean;
}

export interface IVeoFormElementDefaultProps {
  metaData: JSONSchema7;
  disabled: boolean;
  objectCreationDisabled: boolean;
  translations: IBaseObject;
  debug: boolean;
  [key: string]: any;
}

export interface IVeoFormsElementDefinition {
  code: string;
  name: { [lang: string]: string }; // Name of this component (used in FormSchemaEditor)
  description: { [lang: string]: string }; // Description describing this component (used in FormSchemaEditor)
  conditions?: (_props: IVeoFormElementDefaultProps) => boolean[]; // Every condition has to be truthy in order for this component to get displayed. Components with more truthy conditions will be displayed first.
  bias?: number; // Bias so this component gets choosen over other components with the same amount of truthy conditions
}

export interface IVeoFormElementFormSchemaRule {
  effect: 'HIDE' | 'SHOW' | 'DISABLE' | 'ENABLE';
  condition: {
    scope: string;
    schema: JSONSchema7;
  };
}

export interface IVeoFormElementFormSchema {
  type: string;
  options: IBaseObject;
  elements?: IVeoFormElementFormSchema[];
  rule?: IVeoFormElementFormSchemaRule;
}

export interface IVeoFormControlFormSchema extends IVeoFormElementFormSchema {
  scope: string;
  label: string;
}

export interface IVeoFormLabelFormSchema extends IVeoFormElementFormSchema {
  text: string;
}

export interface IVeoFormWidgetFormSchema extends IVeoFormElementFormSchema {
  name: string;
}
