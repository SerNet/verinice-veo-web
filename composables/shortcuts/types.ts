/*
 * verinice.veo web
 * Copyright (C) 2025 Aziz Khalledi
 *
 * This program is free software: you can redistribute it and/or modify it
 * under the terms of the GNU Affero General Public License
 * as published by the Free Software Foundation, either version 3 of the License,
 * or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
 * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License along with this program.
 * If not, see <http://www.gnu.org/licenses/>.
 */

export const SHORTCUT_DIALOG_DISPLAY = 'Shift + ?';
export const CATEGORY_DOMAIN_NAVIGATION = 'Domain Navigation';
export const SEQUENCE_TIMEOUT_MS = 1000;
export const MAX_SEQUENCE_LENGTH = 3;

export interface ShortcutAction {
  id: string;
  name: string;
  keys: string;
  description: string;
  action: () => void;
  category?: string;
  disabled?: boolean;
}

export interface ShortcutGroup {
  id: string;
  name: string;
  shortcuts: ShortcutAction[];
}

export interface TrieNode {
  shortcuts: ShortcutAction[];
  children: Map<string, TrieNode>;
}

export interface MatchResult {
  exactMatches: ShortcutAction[];
  hasPartialMatches: boolean;
}

export interface KeyboardHandlerOptions {
  onExactMatch: (shortcut: ShortcutAction) => void;
  onSequenceReset: () => void;
  isEnabled: () => boolean;
}

export interface ShortcutRegistry {
  register: (group: ShortcutGroup) => void;
  unregister: (id?: string) => void;
  clear: () => void;
}
