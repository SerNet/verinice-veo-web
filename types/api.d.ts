import { JSONSchema6 } from "json-schema";

import { ID_FIELD, PARENT_FIELD, TITLE_FIELD, TYPE_FIELD } from "~/config/api";
export * from "~/config/api";

export type ApiItemFields = typeof ID_FIELD | typeof PARENT_FIELD | typeof TITLE_FIELD | typeof TYPE_FIELD;

export type UUID = string;

export interface ApiItem {
  [ID_FIELD]?: UUID;
  [PARENT_FIELD]: UUID;
  [TITLE_FIELD]: string;
  [TYPE_FIELD]: string;
  schema?: string;
  title?: string;
}

export interface ApiLink {
  //id: string;
  [ID_FIELD]: UUID;
  [TYPE_FIELD]: UUID;
  source: string;
  target: string;
  schema: string;
}

export interface ApiHistory {
  author: string;
  data: string;
  dataId: string;
  timestamp: string;
}

export interface ApiUserTokenPayload {
  exp: number;
  sub: string;
  iss: string;
  iat: number;
  aud: string;
  profiles: string[];
}

export interface ApiSchema extends JSONSchema6 {}

interface ApiRoutes {
  "/login": {
    POST: {
      body: {
        username: string;
        password: string;
      };
    };
  };
  "/elements": {
    GET: {
      query: {
        parent?: UUID;
      };
      response: ApiItem[];
    };
    POST: {
      body: ApiItem;
    };
  };
  "/elements/:id": {
    GET: {
      params: {
        id: UUID;
      };
      response: ApiItem;
    };
    PUT: {
      params: {
        id: UUID;
      };
      body: ApiItem;
    };
    DELETE: {
      params: {
        id: UUID;
      };
    };
  };
  "/elements/:id/children": {
    GET: {
      params: {
        id: UUID;
      };
      response: ApiItem[];
    };
  };
  "/elements/:id/links": {
    GET: {
      params: {
        id: UUID;
      };
      response: ApiLink[];
    };
  };
  "/elements/:id/history": {
    GET: {
      params: {
        id: UUID;
      };
      response: ApiLink[];
    };
  };
  "/links/:id": {
    DELETE: {
      params: {
        id: UUID;
      };
    };
  };
  "/schemas": {
    GET: {
      params: {
        id: UUID;
      };
      response: string[];
    };
  };
  "/schemas/:name.json": {
    GET: {
      params: {
        name: string;
      };
      response: ApiSchema;
    };
  };
}
