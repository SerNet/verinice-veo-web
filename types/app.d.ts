import { ApiItem, UUID } from "~/types/api";

export interface AppElement {
  id: UUID;
  title: string;
  parent?: UUID;
  type: string;
  data: ApiItem;
}

export interface AppLink {
  id: UUID;
  type: string;
  sourceId: UUID;
  targetId: UUID;
  source?: AppElement;
  target?: AppElement;
}

export type AppElementMap = Record<UUID, AppElement>;
export type AppElementsMap = Record<UUID, AppElement[]>;
export type UUIDsMap = Record<UUID, UUID[]>;
