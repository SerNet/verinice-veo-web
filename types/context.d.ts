import { API } from '~/plugins/api'
import { Auth } from '~/plugins/auth'

declare module 'vue/types/vue' {
  interface Vue {
    $api: API
    $auth: Auth
  }
}

declare module '@nuxt/types/app' {
  interface NuxtAppOptions {
    $api: API
    $auth: Auth
  }
}
