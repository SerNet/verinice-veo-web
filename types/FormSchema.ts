import { UISchema } from '~/types/UISchema'

export interface BaseFormSchema {
  name: string
  modelType: string
  subType: string | null
}

export interface FormSchema extends BaseFormSchema {
  content: UISchema
}

export interface FormSchemaMeta extends BaseFormSchema {
  id: string
}

export type FormSchemaMetas = FormSchemaMeta[]
