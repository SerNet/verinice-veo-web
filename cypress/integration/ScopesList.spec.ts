/*
 * verinice.veo web
 * Copyright (C) 2021  Jonas Heitmann
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
/// <reference path="../support/index.d.ts" />
import { IVeoEntity } from '../../types/VeoTypes';

describe('Scopes List', () => {
  beforeEach(() => {
    cy.auth();

    cy.interceptLayoutCalls();
    cy.visit(`/unit-d496f98f-c051-443c-9b1f-65d65b64996d/scopes/-/list/`);
    cy.wait('@G_fetchObjects');
  });

  // Only checking for correct request, we expect the vuetify component to work correctly
  it('Navigates the forms table using the forward/back buttons', function () {
    cy.intercept({
      method: 'GET',
      url: /.*\/api\/(assets|controls|documents|incidents|persons|processes|scenarios|scopes)\?(.+)$/,
      query: { page: '1' }
    }).as('fetchObjects1');

    cy.intercept({
      method: 'GET',
      url: /.*\/api\/(assets|controls|documents|incidents|persons|processes|scenarios|scopes)\?(.+)$/,
      query: { page: '0' }
    }).as('fetchObjects2');

    // Go to next page (from page 0 to page 1)
    cy.get('.v-data-footer__icons-after .mdi-chevron-right').click();
    cy.wait('@fetchObjects1').its('request.url').should('contain', 'page=1');

    // Go to previous page (from page 1 to page 0)
    cy.get('.v-data-footer__icons-before .mdi-chevron-left').click();
    cy.wait('@fetchObjects2').its('request.url').should('contain', 'page=0');
  });

  // Only checking for correct request, we expect the vuetify component to work correctly
  it('Changes page size of the table', function () {
    const newTablePageSize = 25; // Valid option defined in itemsPerPageOptions in VeoFormList.vue

    // Skip to next page
    cy.get('.v-data-footer__icons-after .mdi-chevron-right').click();
    cy.wait('@G_fetchObjects');

    cy.intercept({
      method: 'GET',
      url: /.*\/api\/(assets|controls|documents|incidents|persons|processes|scenarios|scopes)\?(.+)$/,
      query: { page: '0', size: '25' }
    }).as('fetchObjects');

    // Change page size
    cy.get('[data-cy=veo-object-list-objects-table] .v-data-footer .v-input').closest('.v-select').type(`${newTablePageSize}{enter}`);
    cy.wait('@fetchObjects').should((req: any) => {
      expect(req.request.url).to.contain(`size=${newTablePageSize}`); // Expect page size to increase to previously defined value
      expect(req.request.url).to.contain('page=0'); // Expect page to jump back to first page
    });
    cy.get('.v-data-footer__icons-after button').should('be.disabled'); // There are now more items on this page than exist => disable skip button
  });

  // Only checking for correct request, we expect the vuetify component to work correctly
  it('Tests the sorting for every sortable column', function () {
    let descendingSortings = 0;
    let ascendingSortings = 0;
    let totalQueries = 0;
    const sortableColumns = ['designator', 'name', 'updatedAt', 'updatedBy']; // Array containing the property that should be sorted for (NOTE: Column in the VeoFormList.vue file might be named differently), so we make sure only valid properties are sorted

    cy.intercept({
      method: 'GET',
      url: /.*\/api\/(assets|controls|documents|incidents|persons|processes|scenarios|scopes)\?(.+)$/,
      query: { page: '0' }
    }).as('fetchObjects');

    // Test whether the page jumps back to one if the user changes the sort direction
    cy.get('.v-data-footer__icons-after .mdi-chevron-right').click();
    cy.wait('@G_fetchObjects');
    cy.get('[data-cy=veo-object-list-objects-table]').contains('Objektname').click();
    cy.wait('@fetchObjects').its('request.url').should('contain', 'page=0');

    // Iterate over every sortable table header and check if the correct params are sent
    cy.get('[data-cy=veo-object-list-objects-table] th.sortable').each(($element, _index, _$list) => {
      return new Cypress.Promise((resolve) => {
        // Sort asc and desc
        // eslint-disable-next-line no-unused-vars, @typescript-eslint/no-unused-vars
        for (const sortOrder of ['asc', 'desc']) {
          cy.wrap($element).click();
          cy.get('@G_fetchObjects').should((req: any) => {
            const sortByPropertyExists = sortableColumns.includes(new URL(req.request.url).searchParams.get('sortBy'));
            expect(req.request.url).to.contain(`sortBy=`);
            // eslint-disable-next-line no-unused-expressions
            expect(sortByPropertyExists).to.be.true; // Expect parameter to sort by to be part of sortableColumns
            if (req.request.url.includes('sortOrder=asc')) {
              ascendingSortings++;
            } else if (req.request.url.includes('sortOrder=desc')) {
              descendingSortings++;
            }
            totalQueries++;
          });
        }

        // eslint-disable-next-line no-unused-expressions
        expect(descendingSortings === ascendingSortings).to.be.true;
        // eslint-disable-next-line no-unused-expressions
        expect(totalQueries === descendingSortings * 2).to.be.true;
        resolve();
      });
    });
  });

  // Checks whether the list redirects to the correct object for editing
  it('Redirects the user to the edit page of an object', function () {
    const index = 4;
    let id: string;

    cy.wait('@G_fetchObjects');

    // Load fixture to get the id
    cy.fixture('api/default/entities/scopes/fetchAll.json').then((allScopes) => {
      id = allScopes.items[index].id;
    });

    // Click on the edit button
    cy.get('[data-cy=veo-object-list-objects-table] tbody tr:not(.v-data-table__empty-wrapper)')
      .eq(index)
      .within(() => {
        cy.get('[data-cy=veo-object-list-edit-item]').click();
        cy.location('pathname').should('contain', `scope-${id}`);
      });
  });

  // Checks whether the correct url gets called for deleting the object
  it('Deletes an object from the forms table', function () {
    const index = 4;
    let id: string;

    cy.wait('@G_fetchObjects');

    // Load fixture to get the id
    cy.fixture('api/default/entities/scopes/fetchAll.json').then((allScopes) => {
      id = allScopes.items[index].id;
    });

    cy.intercept(
      {
        method: 'DELETE',
        url: /.*\/api\/(assets|controls|documents|incidents|persons|processes|scenarios|scopes)\/(.+)$/
      },
      (req) => {
        expect(req.url).to.contain(id);

        req.reply({
          success: true,
          resourceId: 'string',
          message: 'string'
        });
      }
    );

    // Click on the edit button
    cy.get('[data-cy=veo-object-list-objects-table] tbody tr:not(.v-data-table__empty-wrapper)')
      .eq(index)
      .within(() => {
        cy.get('[data-cy=veo-object-list-delete-item]').click();
      });
    cy.get('[data-cy=veo-delete-entity-dialog-confirm-button]').click();
  });

  it('Clones the object in the forms table', function () {
    const index = 4;
    let entity: IVeoEntity;
    let returnedEntity: IVeoEntity;

    cy.wait('@G_fetchObjects');

    // Load fixture to get the id
    cy.fixture('api/default/entities/scopes/fetchAll.json').then((allScopes) => {
      entity = allScopes.items[index];
      entity.name = `${entity.name} (Klon)`;
      returnedEntity = allScopes;
      delete entity.id;
      delete entity.designator;
    });

    cy.intercept(
      {
        method: 'POST',
        url: /.*\/api\/(assets|controls|documents|incidents|persons|processes|scenarios|scopes)$/
      },
      (req) => {
        expect(req.body).to.deep.equal(entity);

        req.reply(returnedEntity);
      }
    );

    // Click on the edit button
    cy.get('[data-cy=veo-object-list-objects-table] tbody  tr:not(.v-data-table__empty-wrapper)')
      .eq(index)
      .within(() => {
        cy.get('[data-cy=veo-object-list-clone-item]').click();
      });
  });

  it('Views the child items of an entity not containing child entities', function () {
    const index = 4;
    let id: string;

    cy.intercept({
      method: 'GET',
      url: /.*\/api\/(assets|controls|documents|incidents|persons|processes|scenarios|scopes)\?(.+)$/
    }).as('fetchObjects');

    cy.wait('@G_fetchObjects');

    // Load fixture to get the id
    cy.fixture('api/default/entities/scopes/fetchAll.json').then((allScopes) => {
      id = allScopes.items[index].id;
    });

    // Click on the column to go into
    cy.get('[data-cy=veo-object-list-objects-table] tbody tr:not(.v-data-table__empty-wrapper)')
      .eq(index)
      .click()
      .then(() => {
        cy.location('pathname').should('contain', `unit-d496f98f-c051-443c-9b1f-65d65b64996d/scopes/scope-${id}/list`);

        cy.wait('@fetchObjects');
        cy.get('[data-cy=veo-object-list-objects-table] tbody tr:not(.v-data-table__empty-wrapper)').should('have.length', 1);
        cy.get('[data-cy=veo-object-list-navigate-parent]').should('exist');
      });
  });

  it('Views the child items of an entity containing child entities', function () {
    const index = 3;
    let id: string;

    cy.intercept({
      method: 'GET',
      url: /.*\/api\/(assets|controls|documents|incidents|persons|processes|scenarios|scopes)\?(.+)$/
    }).as('fetchObjects');

    cy.wait('@G_fetchObjects');

    // Load fixture to get the id
    cy.fixture('api/default/entities/scopes/fetchAll.json').then((allScopes) => {
      id = allScopes.items[index].id;
    });

    // Click on the column to go into
    cy.get('[data-cy=veo-object-list-objects-table] tbody tr:not(.v-data-table__empty-wrapper)')
      .eq(index)
      .click()
      .then(() => {
        cy.location('pathname').should('contain', `unit-d496f98f-c051-443c-9b1f-65d65b64996d/scopes/scope-${id}/list`);

        cy.wait('@fetchObjects');
        cy.get('[data-cy=veo-object-list-objects-table] tbody tr:not(.v-data-table__empty-wrapper)').should('have.length', 5);
      });
  });

  it('Unlinks a child item from its parent', function () {
    const index = 3;
    const childIndex = 1;

    cy.intercept(
      {
        method: 'PUT',
        url: /.*\/api\/(assets|controls|documents|incidents|persons|processes|scenarios|scopes)\/(.*)$/
      },
      (req) => {
        req.reply(req.body);
      }
    ).as('saveEntity');

    cy.wait('@G_fetchObjects');

    // Click on the column to go into
    cy.get('[data-cy=veo-object-list-objects-table] tbody tr:not(.v-data-table__empty-wrapper)')
      .eq(index)
      .click()
      .then(() => {
        cy.wait(['@G_fetchObjects', '@G_fetchObject']);
        cy.get('[data-cy=veo-object-list-objects-table] tbody tr:not(.v-data-table__empty-wrapper)')
          .eq(childIndex)
          .within(() => {
            cy.get('[data-cy=veo-object-list-unlink-item]').click();
          });
      });
    cy.get('[data-cy=veo-unlink-entity-dialog-confirm-button]').click();
    cy.wait('@saveEntity').its('request.body').toMatchSnapshot();
  });
});
