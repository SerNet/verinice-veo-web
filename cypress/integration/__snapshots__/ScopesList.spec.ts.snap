exports[`Scopes List > Unlinks a child item from its parent #0`] =
{
  "abbreviation": "VS1",
  "createdAt": "2021-07-12T21:43:20.188792Z",
  "createdBy": "pb",
  "customAspects": {},
  "description": "Scope mit Subtype Verantwortliche Stelle",
  "designator": "SCP-7",
  "displayName": "SCP-7 VS1 Brasilien",
  "domains": [
    {
      "displayName": "DSGVO DSGVO",
      "resourcesUri": "https://veo.develop.cpmsys.io/domains",
      "searchesUri": "https://veo.develop.cpmsys.io/domains/searches",
      "targetUri": "https://veo.develop.cpmsys.io/domains/ed67e4d7-c657-4479-ba8a-c53999d2930a"
    }
  ],
  "id": "5986694a-c58a-4411-8991-9fc34845bc6a",
  "links": {},
  "members": [
    {
      "displayName": "PRO-10 P4 Cyan",
      "resourcesUri": "https://veo.develop.cpmsys.io/processes{?unit,displayName,subType,size,page,sortBy,sortOrder}",
      "searchesUri": "https://veo.develop.cpmsys.io/processes/searches",
      "targetUri": "https://veo.develop.cpmsys.io/processes/f1c36cf9-70ab-4726-a993-2c23fddcad68"
    },
    {
      "displayName": "AST-6 DA1 Dachs",
      "resourcesUri": "https://veo.develop.cpmsys.io/assets{?unit,displayName,subType,size,page,sortBy,sortOrder}",
      "searchesUri": "https://veo.develop.cpmsys.io/assets/searches",
      "targetUri": "https://veo.develop.cpmsys.io/assets/a2304df4-8126-4709-ba70-53b4d34c3320"
    },
    {
      "displayName": "PRO-9 P3 Blau",
      "resourcesUri": "https://veo.develop.cpmsys.io/processes{?unit,displayName,subType,size,page,sortBy,sortOrder}",
      "searchesUri": "https://veo.develop.cpmsys.io/processes/searches",
      "targetUri": "https://veo.develop.cpmsys.io/processes/fc10e550-8f5b-4d31-9bdb-b4b898a9d254"
    }
  ],
  "name": "Brasilien",
  "owner": {
    "displayName": "NBRX Testunit",
    "resourcesUri": "https://veo.develop.cpmsys.io/units{?parent,displayName}",
    "searchesUri": "https://veo.develop.cpmsys.io/units/searches",
    "targetUri": "https://veo.develop.cpmsys.io/units/d496f98f-c051-443c-9b1f-65d65b64996d"
  },
  "subType": {
    "ed67e4d7-c657-4479-ba8a-c53999d2930a": "SCP_ResponsibleBody"
  },
  "type": "scope",
  "updatedAt": "2021-07-12T21:43:20.188792Z",
  "updatedBy": "pb"
};
